docs/design/logging-integration.adoc

```diff
@@ -0,0 +1,196 @@
+This is all configured elsewhere and outside of KubeArchive.
+- When the `Pod` is archived, the URL for accessing the log should be generated
+and stored with it. There is no attempt to query the logging system to verify
+the existence of the log.
+
+====
+
+Here's another sample `KubeArchiveConfig` for `PipelineRuns`:
+```
+---
+apiVersion: kubearchive.kubearchive.org/v1alpha1
+kind: KubeArchiveConfig
+metadata:
+  name: kubearchive
+  namespace: test
+spec:
+  resources:
+    - selector:
+        apiVersion: tekton.dev/v1
+        kind: PipelineRun
+      deleteWhen: has(status.completionTime)
+    - selector:
+        apiVersion: tekton.dev/v1
+        kind: TaskRun
+      archiveOnDelete: true
+    - selector:
+        apiVersion: v1
+        kind: Pod
+      archiveOnDelete: has(body.metadata.labels["tekton.dev/pipeline"])
+```
+In this case the following happens:
+
+- `PipelineRuns` are archived when they complete.
+- `TaskRuns` are archived when they are deleted.
+- `Pods` are archived when they are deleted and are also part of a Tekton Pipeline
+
+== Generating Log URLs
+
+The logging system is an integration in KubeArchive. All that is required for
+the integration is the successful generation of a URL to access the log for a
+specific `Pod` in the logging system. URLs will ONLY be generated for `Pods`.
+
+To support multiple logging systems, the URLs must be able to be parameterized
+based on the logging system. This could be done via a `ConfigMap` named for the
+logging system being implemented. This `ConfigMap` requires a single entry,
+LOG_URL, whose string value will be evaluated in an environment where variables
+in it will be substituted. For example, take this `ConfigMap` for Splunk:
+```
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: kubearchive-splunk
+  namespace: kubearchive
+data:
+  LOG_URL=https://splunk-host/app/search/search?q=search *| spath "kubernetes.pod_id" | search "kubernetes.pod_id"="${POD_ID}"
+  POD_ID=".body.pod_id"
+```
+Any other entry in this `ConfigMap` is treated as a variable, and the value of
+each variable is a JSON path expression that is evaluated against the body of
+the cloud event (i.e. the resource) to determine the real value of that variable
+used in the substitution. For example:
+
+So when generating the logging URL to be stored when a `Pod` is archived,
+the following steps are done:
+
+1. An "environment" is created by evaluating the JSON path expression in each entry
+   in the `ConfigMap`, excluding LOG_URL.
+1. The value for LOG_URL is then evaluated using this environment, resulting in the
+   final URL to the log in the logging system.
+
+== Implementation
+
+=== Database
+
+The KubeArchive database will have a table named `log_url` with three fields:
+
+1. An `id` field as a primary key.
+1. A `resource_id` field which is a foreign key to `resource.id`.  
+1. A `url` field which is the URL for one of of logs.
```

Comment:
There is a typo here:
```suggestion
1. A `url` field which is the URL for one of the logs.
```

BTW, I have a question here, let's say we have a user that uses splunk for example, so we have un our database the path to the specific URL to find the logs of the pod, something like this: `http://<the-splunk-host>:<the-splunk-port>/<path>?<queryfilter>`

Now, let's say that the user need to update something about the URL, can be the host, or maybe the port, or maybe even the protocol, then, all the stored URLs will be outdated and stop working.

I think, that it may be more useful to build the URL dynamically based on the content of the `ConfigMap` that stores the URL.

Reply:
Typo is fixed.

Reply:
The more I think about this the more I am think that generating the log URL at archive time is the way to do it.  We've come up with a number of ways these URLs can be broken based on settings.

- The hostname changes
- The configuration of the logging system changes.
- The setting for LOG_URL or other variables in the `ConfigMap` changes.

These can all break it.  I conclude that there's no perfect solution here.  Even splitting the "prefix" as you suggest isn't perfect either. It's broken just as easily as anything else., and here's one scenario:

What if the there is new logging host, but the old logging host is still around, available, and contains the old logs?

In this scenario all the old logging links would be broken if we allowed the "prefix" to change.

I think the safest thing, and the best we can do, is to generate the logs at archive time and store them in the `log_url` table. It also keeps things simple as the log URL generation is all done in the sink, instead of being split between the sink and the API.

If there are changes that make the log URLs incorrect, admins have other ways to fix them.  They can write SQL queries to update the URLs with the appropriate changes. Or, in the case of a hostname change as you mentioned, then can either do that or provide a DNS entry to point to the new server.  There are lots of different solutions, but this problem is outside of KubeArchive.

Reply:
I agree, you have made a fair point and I am conviced of storing it the way you suggest, on "archiving" time. BTW, I think that this is a concern that might be brought up in the future so we should write this decission somewhere in the docs.

Reply:
@maruiz93, I just realized this is already mentioned in this document and flagged as important:
```
[IMPORTANT]
====
- KubeArchive has no responsibility for sending the logs to the logging system.
This is all configured elsewhere and outside of KubeArchive.
- When the `Pod` is archived, the URL for accessing the log should be generated
and stored with it. There is no attempt to query the logging system to verify
the existence of the log.
====
```
But it should probably also be included in any user documentation that is written for logging.